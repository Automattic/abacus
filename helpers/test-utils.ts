import mediaQuery from 'css-mediaquery'

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}

export { createMatchMedia }
