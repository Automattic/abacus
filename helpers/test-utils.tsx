import { Queries, render as actualRender, RenderOptions } from '@testing-library/react'
import mediaQuery from 'css-mediaquery'
import React from 'react'

import ThemeProvider from '@/styles/ThemeProvider'

/**
 * A wrapped unit-test react-renderer, useful for adding React Contexts globally.
 */
export const render: typeof actualRender = <Q extends Queries>(ui: React.ReactElement, options?: RenderOptions<Q>) =>
  actualRender((<ThemeProvider>{ui}</ThemeProvider>) as React.ReactElement, options) as ReturnType<typeof actualRender>

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
export function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}
