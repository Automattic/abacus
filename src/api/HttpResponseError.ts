import { WretcherError as WretcherError, WretcherResponse } from 'wretch'

/**
 * A HTTP Error we receive from the server.
 * 
 * WretcherError gets close but isn't friendly enough.
 */
export default class HttpResponseError extends Error implements WretcherError {
    public status: number
    public response: WretcherResponse
    public text?: string
    public json?: Object
    public name = 'HttpResponseError'

    /**
     * This constructor is only for mocking.
     * @param status 
     */
    constructor(status: number, ...vendorSpecificParams: unknown[]) {
        // @ts-ignore vendorSpecificParams
        super(`${status}`, ...vendorSpecificParams)
        this.status = status
    }
}

/**
 * Transform a WretcherError to HttpResponseError
 * 
 * Only necessary as a WretcherError isn't a named error :/
 * 
 * @param wretcherError 
 */
export function wretcherErrorToHttpResponseError(wretcherError: WretcherError): HttpResponseError {
    wretcherError.name = 'HttpResponseError'
    wretcherError.message = `${wretcherError.status} ${wretcherError.response.statusText}`
    Object.setPrototypeOf(wretcherError, HttpResponseError.prototype)
    // istanbul ignore next; V8 engine only
    if (Error.captureStackTrace) {
        Error.captureStackTrace(wretcherError, HttpResponseError);
    }
    return wretcherError
}
