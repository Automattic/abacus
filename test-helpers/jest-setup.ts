import { format } from 'util'

// Throw errors on any console logging.
// Adapted from https://github.com/facebook/jest/issues/6121#issuecomment-529591574
;['error', 'info', 'log', 'warn'].forEach((funcName) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(global.console as any)[funcName] = function (message?: any, ...optionalParams: any[]) {
    throw new Error(format(message, optionalParams))
  }
})
