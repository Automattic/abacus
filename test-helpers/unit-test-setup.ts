/* eslint-disable @typescript-eslint/ban-ts-ignore */
// @ts-ignore
window.URL.createObjectURL = jest.fn()

// @ts-ignore
window.HTMLCanvasElement.prototype.getContext = jest.fn()
