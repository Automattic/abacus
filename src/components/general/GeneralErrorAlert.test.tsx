import { StatusCodes } from 'http-status-codes'
import React from 'react'

import HttpResponseError from 'src/api/HttpResponseError'
import { render } from 'src/test-helpers/test-utils'

import GeneralErrorAlert from './GeneralErrorAlert'

test('renders nothing for no error', () => {
  const { container } = render(<GeneralErrorAlert />)
  expect(container).toMatchSnapshot()
})

test('renders an HttpResponseError with server message in JSON', () => {
  const error = new HttpResponseError(StatusCodes.BAD_REQUEST)
  error.response = new Response(
    '{"code":"invalid_name","message":"The experiment name is already taken","data":{"status":400}}',
  )
  error.json = { code: 'invalid_name', message: 'The experiment name is already taken', data: { status: 400 } }
  const { container } = render(<GeneralErrorAlert error={error} />)
  expect(container).toMatchSnapshot()
})

test('renders an HttpResponseError without server message in JSON', () => {
  const error = new HttpResponseError(StatusCodes.BAD_REQUEST)
  const { container } = render(<GeneralErrorAlert error={error} />)
  expect(container).toMatchSnapshot()
})

test('renders an generic Error', () => {
  const error = new Error('This is a generic error')
  const { container } = render(<GeneralErrorAlert error={error} />)
  expect(container).toMatchSnapshot()
})
