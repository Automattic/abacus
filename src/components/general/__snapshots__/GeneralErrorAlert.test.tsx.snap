// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an HttpResponseError with server message in JSON 1`] = `
<div>
  <div
    class="MuiPaper-root MuiAlert-root MuiAlert-standardError makeStyles-root-2 MuiPaper-elevation0"
    role="alert"
  >
    <div
      class="MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message"
    >
      <div
        class="MuiTypography-root MuiAlertTitle-root MuiTypography-body1 MuiTypography-gutterBottom"
      >
        Error Response: 
        400
         
        
      </div>
      The experiment name is already taken
    </div>
  </div>
</div>
`;

exports[`renders an HttpResponseError without server message in JSON 1`] = `
<div>
  <div
    class="MuiPaper-root MuiAlert-root MuiAlert-standardError makeStyles-root-3 MuiPaper-elevation0"
    role="alert"
  >
    <div
      class="MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message"
    >
      <div
        class="MuiTypography-root MuiAlertTitle-root MuiTypography-body1 MuiTypography-gutterBottom"
      >
        Error Response: 
        400
         
        
      </div>
    </div>
  </div>
</div>
`;

exports[`renders an generic Error 1`] = `
<div>
  <div
    class="MuiPaper-root MuiAlert-root MuiAlert-standardError makeStyles-root-4 MuiPaper-elevation0"
    role="alert"
  >
    <div
      class="MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message"
    >
      <div
        class="MuiTypography-root MuiAlertTitle-root MuiTypography-body1 MuiTypography-gutterBottom"
      >
        Error
        :
      </div>
      This is a generic error
    </div>
  </div>
</div>
`;

exports[`renders nothing for no error 1`] = `<div />`;
